### PR TITLE
Add unified trainer (PC + backprop + autoregressive)

### PR DESCRIPTION
**Pull request by aprotsenko24**
## What

One self-contained, algorithm-parameterized trainer that consolidates the three legacy trainers (`train.py` / `train_backprop.py` / `train_autoregressive.py`) behind a single API:

```python
train(params, structure, loader, optimizer, config, key,
      algorithm="pc"|"backprop", autoregressive=False, ...)
evaluate(..., algorithm=..., autoregressive=False, ...)
```

## Design

- The two learning rules differ in **exactly three funnels** — `_forward` / `compute_gradients` / `_eval_step`. Everything else (batch conversion, clamping, optimizer update, the epoch/batch driver, sharding, eval reduction) is shared.
- **Autoregressive is an orthogonal boolean flag**, not a third algorithm: its only footprint is a causal-mask clamp + loss/perplexity reporting; it works with either algorithm.
- The gradient math is called **verbatim** from `fabricpc.core` (`compute_local_weight_gradients`, `run_inference`, `initialize_graph_state`) — not reimplemented.
- The three legacy trainer files are **left untouched** as the parity baseline (this is a new, standalone file).
- `algorithm` / `autoregressive` / `use_causal_mask` / `metric_name` are explicit signature parameters (not config keys); `config` holds only `num_epochs`.

## Numerical parity (single device)

- `scripts/_parity_check_unified.py` — PC & backprop, train + eval vs the legacy trainers, **exact to 1e-6**.
- `scripts/_parity_check_ar.py` — AR-PC / AR-backprop vs legacy: **eager gradients bit-identical (atol=0)**; jitted params ≤2e-3 (a float artifact — the *legacy* AR step computes an unused cross-entropy inside its jit, which the unified version omits); includes a ragged-batch eval assertion.
- `tests/test_unified_trainer.py` — PC & backprop train + eval parity, **1e-12** (pytest gate).

All green on CPU (`JAX_PLATFORMS=cpu`).

## Known AR caveat (do not "fix")

AR eager gradients are bit-identical to the legacy formula; only the *jitted* params drift ≤2e-3, purely because the legacy AR step computes a dead cross-entropy inside the jit (changes XLA float-scheduling; Adam amplifies it). The unified step is cleaner and run-to-run deterministic.

## Deferred (separate atomic step, after sign-off)

Wiring into `training/__init__.py` and converting the three legacy files to deprecation shims that delegate to this module.

### CODE REVIEW

**Comment by matthewbehrend** [comment]
Design is substantially revised. Added planning file.
Start here: docs/dev_plans/unified_trainer_api.md for design and verification results.
Old training harnesses removed with full migration to fabricpc/training/trainer.py

**Review by matthewbehrend — COMMENTED** [review]
Thank you Artem for this PR. It makes life much easier in training models and to maintain going forward. Here's what worked in the design as submitted and was retained or developed further:

The prototype unified trainer structural skeleton survived intact; what HEAD replaced was its parameter surface and the legacy machinery it had inherited.

1. **One step body, one loop for both algorithms.** The prototype built exactly one step body, not two, with a single `run_training_loop` and `_make_step` closing over static config. HEAD even shrank the branch: `_batch_grads` is now the only place the two algorithms diverge, because clamping and the objective became fully shared.

2. **`algorithm` as an explicit signature parameter, not a config key.** The prototype's docstring argued the choice should be visible at the call site rather than buried in a dict. Retained verbatim, tightened to a `Literal`, and extended to `generate`. This is also what made the `functools.partial(train, algorithm="backprop")` composition with `ExperimentArm` possible.

3. **Build-time validation of prerequisites encoded in the graph.** `_validate_algo` — PC requires an inference algorithm in `structure.config`, backprop requires `FeedforwardStateInit` — carried over as `_validate_algorithm` with the same rules, now also imported by `generation.py`. HEAD generalized the fail-fast stance it established: required `num_epochs`, retired config keys raise, target shapes validated, loader `len()` checked with an actionable message.

4. **Differentiating through the feedforward initializer.** The prototype's backprop path already took `jax.value_and_grad` through `FeedforwardStateInit`, proving the graph's own forward pass is differentiable end to end. HEAD's energy framing is built directly on that mechanism — it kept the differentiation path and deleted the loss-selection layer (`_metric_cross_entropy`, `_metric_mse`) that sat beside it, replacing the objective with `graph_energy` over the clamped target nodes.

5. **`build_clamps` with `clamp_target` as the train/eval switch.** One clamp builder mapping task keys through `structure.task_map`, with causal-mask injection as "the only graph change autoregressive mode makes" — that observation is what made the `autoregressive` flag deletable. HEAD kept the function and completed the idea: the mask trigger moved from a caller flag to the `TaskMap` itself, sequence length now reads from the mask node's declared shape rather than `batch["x"]`, and one-hot plus shape validation were added.

6. **The custom-metric escape hatch.** The prototype's `metric_fn(state, batch, structure) -> scalar` fixed the right argument signature — metric computed on the settled state inside the step, no extra forward work. `EvalMetric.fn` keeps exactly those arguments; HEAD developed the return contract to `(value, weight)` plus `finalize`, which fixed the aggregation (weighted, not per-batch mean), the one-metric-per-call limit, and post-aggregation transforms like perplexity.

7. **Callback return replaces the stored history entry.** Already present in the prototype loop (`batch_metrics.append(iter_callback(...) if iter_callback else metric)`; `epoch_results.append(epoch_callback(...))`). HEAD kept the rule and fixed its sharp edge: only a non-None return replaces the entry, so a logging-only callback no longer erases the stored metrics. Likewise, callbacks were unguarded at 772795b by omission; HEAD elevated exception propagation to a tested contract that tuner pruning depends on.

8. **Loop ergonomics carried over nearly unchanged.** Fractional epochs (`frac = num_epochs - floor`), the ragged-batch skip-with-one-time-warning policy, tqdm with per-batch metric postfix, the `(params, structure, loader, optimizer, config, rng_key)` positional prefix, and `convert_batch`/`create_causal_mask` all survive at HEAD — refined (zero-batch tails dropped, `use_tqdm` folded into `verbose`, everything after `rng_key` keyword-only) rather than redesigned.

9. **The internal-energy distinction.** The prototype's `_metric_energy(internal_only=...)` already separated the `in_degree>0` training objective from an all-node sum. HEAD generalized it into `graph_energy(node_names=)` with the `in_degree>0` default and unified eval onto the same definition.

What did not survive is consistent with this picture: the prototype's weak points were its inherited machinery (pmap/pmean, the split-chain RNG, the discarded `opt_state`, the 3-tuple return) and its parameter surface (`autoregressive`/`use_causal_mask`/`metric_name`). Architecture was strong and the unification skeleton, the graph-encoded validation, and the shared-clamping insight all proved out and were kept.

What changed after the design review:
Each item states that baseline prototype design and what replaced it at HEAD.

1. **Public surface.** Retired the legacy trainers. Exported the unified tool set — `train`, `evaluate`, `make_train_step`, `generate`, `build_clamps`, `convert_batch`, `create_causal_mask`, `TrainResult`, `EpochContext`, `EvalMetric`, the `metrics` submodule. All legacy modules are deleted and every caller (examples, tuner, experiments, dashboarding, tests) is migrated; no shims.

2. **Mode axes.** Was: the prototype took `algorithm=` plus `autoregressive=` and `use_causal_mask=` flags, and `_require_causal_mask_node` raised on graphs without a mask node. Now: one axis, `algorithm ∈ ("pc", "backprop")`. The causal mask is injected only when the graph's `TaskMap` declares a mask node (sequence length from that node's declared shape); one-hot encoding follows the target dtype (class count from the target node's `shape[-1]`). The two flags and the mask-node requirement are gone.

3. **Objective.** Was: the training objective was selected by `metric_name`/`metric_fn` (`"energy"`, `"cross_entropy"`, `"mse"`) — a loss-selection layer beside the graph. Now: the objective is energy in both algorithms (`_batch_grads`, trainer.py:284): PC settles latents via `run_inference` and takes local gradients on `graph_energy` over all `in_degree>0` nodes; backprop differentiates `graph_energy` over the clamped target nodes through the feedforward initialization with `jax.value_and_grad`. The output node's energy functional therefore selects the loss; the metric-selection machinery is deleted and the retired `loss_type`/`use_causal_mask` config keys raise.

4. **Return type and resumability.** Was: `train` returned `(params, iter_results, epoch_results)`; `opt_state = optimizer.init(params)` was created internally and discarded, so optimizer moments and schedule counts could not survive a save/load boundary. Now: `TrainResult(params, opt_state, step, iter_results, epoch_results)`, with `opt_state=` and `start_epoch=` accepted for resume.

5. **RNG stream.** Was: a split chain — `split(rng_key)` per epoch, `split(epoch_key, max_batches)` per batch — coupling keys to loader length and making resume require replaying splits. Now: `fold_in(rng_key, epoch_idx)` → `fold_in(epoch_key, batch_idx)`, a pure function of the indices; an interrupted-then-resumed run is bitwise equal to the uninterrupted one.

6. **Device parallelism.** Was: `jax.pmap` with `lax.pmean` on gradients (a per-device mean), parameter replication, `shard_batch`, and a `pmap_single_device` flag; params de-replicated before callbacks. Now: one jitted step for both cases, with an optional `mesh=` placing batches as `NamedSharding(mesh, P("data"))` and params replicated (`"model"` axis reserved). Gradients become the global batch sum, matching the single-device path. All replication/sharding helpers are deleted.

7. **Eval metrics.** Was: one metric per `evaluate` call (`metric_fn`/`metric_name`), per-batch mean aggregation, and a single result key resolved from the algorithm (`"energy"`/`"loss"`). Now: a dict of named `EvalMetric(fn, finalize)` entries with a per-sample `(value, weight)` contract; aggregation is `finalize(Σvalue/Σweight)` across batches and devices, so ragged batches and per-token metrics weight correctly. `metrics=None` selects graph-derived defaults keyed off the target node's energy functional (`cross_entropy`/`perplexity` only for `CrossEntropyEnergy` targets). The `loss` key is renamed `cross_entropy`.

8. **Callback contracts.** Was: `epoch_callback(epoch_idx, params, structure, config, rng_key)` — five positionals with no optimizer state — and `iter_callback(epoch_idx, batch_idx, metric: float)`, with a per-batch `float()` sync on every step. Now: `epoch_callback(ctx: EpochContext)` (adds `step`, `opt_state`, `metrics`; grows by field addition) and `iter_callback(epoch_idx, batch_idx, metrics: dict)`; metrics stay device scalars until epoch boundaries unless a callback forces the sync. Exceptions propagate; a non-None return replaces the stored history entry.

9. **Custom-loop step.** Was: the per-batch step was private (`_make_step`/`compute_gradients`). Now: `make_train_step(structure, optimizer, algorithm=, mesh=)` is public, returns `(params, opt_state, metrics, final_state)`, and does not donate inputs; the internal loop uses a donating variant that drops `final_state` to cut peak memory.

10. **Generation.** Was: `generate_autoregressive` always settled via `run_inference`. Now: `generate(..., algorithm=)` shares `_validate_algorithm` and `build_clamps` with the trainer; `"backprop"` samples from the feedforward pass, required for graphs built without an inference object.

11. **Energy primitive.** Was: the graph energy sum was open-coded in each module. Now: `graph_energy(state, structure, node_names=)` in `fabricpc/core/energy.py`, defaulting to all `in_degree>0` nodes; the trainer, metrics, and dashboarding all call it.

12. **Fail-fast over silent defaults.** Was: `config.get("num_epochs", 10)`; wrong-shape targets surfaced as XLA broadcast errors; no-target eval silently returned zeros. Now: a missing `num_epochs` raises, `build_clamps` shape-validates targets with an actionable `ValueError`, a mesh without a `"data"` axis raises, and the default eval metrics raise on a graph with no target task key. `use_tqdm` is folded into `verbose`.

Notable moves from the prototype unification are items 2, 3, 4, 7, and 8 — the rest replace legacy-harness design that the prototype had carried over (pmap, split-chain RNG, 3-tuple return).

Full detail: `CHANGELOG.md` `[0.5.0]` (migration table, behavior notes) and `docs/dev_plans/unified_trainer_api.md` (rationale, alternatives considered, verification results).

**Review by quinbez — APPROVED** [review]
This is a well executed consolidation. I went through the plan doc and then checked the implementation against it in detail, `trainer.py` (`train`, `evaluate`, `build_clamps`, `make_train_step`), `metrics.py`'s pluggable `EvalMetric` system, the new `graph_energy` function in `core/energy.py`, and `generation.py`'s sampling loop. The core design decision, framing backprop as energy minimization on the same graph structures PC uses, rather than bolting on a separate loss-based path, is carried through consistently across the whole diff, and the trickier pieces (the RNG `fold_in` scheme for bitwise-reproducible resume, the cyclic-graph fix in `graph_energy`'s node ordering, the mesh/sharding logic replacing the old `pmap` path) all match what's documented. Notably, the plan itself calls out nine gaps it found in its own review pass (RNG-test vacuity, a CI coverage regression from deleting `test_multi_gpu.py`, two fragile pseudocode derivations, a missing algorithm axis on `generate()`, among others), I checked each one against the actual code and confirmed all nine are correctly fixed.

Test coverage backs this up well: parity is verified against hand-rolled reference implementations at `1e-12` tolerance for PC and against `jax.grad` for backprop, resume is proven bitwise-identical across a split vs. non-split run, and the new `test_sharding.py` covers single-device/multi-device parity, ragged-batch handling, and padded-sample weighting, closing the gap left by the deleted multi-GPU test file, with the CI workflow updated to actually exercise it via `XLA_FLAGS`. The mechanical migrations across the example scripts and user guides are consistent with the new API throughout. Approving, this is a thorough, and a clear improvement over the four duplicated trainers it replaces.
